src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -463,10 +463,12 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> DumpVisitor<'l, 'tcx, 'll, O> {
         }
 
         // walk type and init value
-        self.visit_ty(typ);
-        if let Some(expr) = expr {
-            self.visit_expr(expr);
-        }
+        self.nest_tables(id, |v| {
```

Comment:
`nest_tables` is used inconsistently in this file. The type of constant and static items is inside the `nest_tables` closure, while the return type for function items is outside. I don't think it's a problem for function/const/static items, since they all have their own generics, but e.g. an associated constant's type can depend on associated types.

I think the `visit_ty` should be outside the `nest_tables`, but I'm not sure it changes anything. The following test should expose any lingering problems I think.

```rust
pub fn g() {
    trait Sup {
        type Assoc;
    }
    trait Trait: Sup {}
    impl Trait {
        const FLAG: <Self as Sup>::Assoc = bogus.field;
    }
}
```

alternatively try


```rust
pub fn g() {
    trait Trait {}
    impl<T> Trait {
        const FLAG: T = bogus.field;
    }
}
```

Reply:
@oli-obk when `visit_ty` was outside `nest_tables`, the snippet in https://github.com/rust-lang/rust/issues/59134#issuecomment-490655542 still ICE'd, so I believe we should nest the tables when visiting assoc const's type as well?

Reply:
Hmm... I just looked through this file and couldn't find any place where trait items are wrapped in `nest_tables` with the id of the impl block or trait block. I think https://github.com/rust-lang/rust/blob/bf36208a10f5bd06a2cf3f35d8f92e43e3ec4fb3/src/librustc_save_analysis/dump_visitor.rs#L765-L768 should have a `nest_tables` with the id of the impl/trait block. The same for https://github.com/rust-lang/rust/blob/bf36208a10f5bd06a2cf3f35d8f92e43e3ec4fb3/src/librustc_save_analysis/dump_visitor.rs#L684-L687

Once that has been done, it should be possible to just wrap the constant body in `nest_tables` here